### PR TITLE
Push xDS when target ports are changed

### PR DESCRIPTION
**Pull request by dddddai**
**Please provide a description of this PR:**
Fixes https://github.com/istio/istio/issues/48580

### CODE REVIEW

File: pilot/pkg/serviceregistry/kube/controller/util.go [hunk 1]
```diff
@@ -89,30 +89,6 @@ func findPortFromMetadata(svcPort v1.ServicePort, podPorts []model.PodPort) (int
...
-			if p.TargetPort.Type == intstr.Int && p.TargetPort.IntVal > 0 {
-				return serviceTargetPort{num: int(p.TargetPort.IntVal), name: p.Name, explicitName: false}
-			}
-			return serviceTargetPort{num: 0, name: p.TargetPort.StrVal, explicitName: true}
-		}
-	}
-	// should never happen
-	log.Debugf("did not find matching target port for %v on service %s", servicePort, k8sService.Name)
-	return serviceTargetPort{num: 0, name: "", explicitName: false}
-}
```
Comment (bleggett) [inline pilot/pkg/serviceregistry/kube/controller/util.go:114]: IIRC some of this is to handle cases where the targetPort is a name _or_ a number?

And that check is lost with the changes?


Are we sure that's desired?

File: pilot/pkg/serviceregistry/serviceentry/conversion_test.go [hunk 2]
```diff
@@ -649,101 +702,101 @@ func testConvertServiceBody(t *testing.T) {
 			externalSvc: httpNone,
 			services: []*model.Service{
 				makeService("*.google.com", "httpNone", constants.UnspecifiedIP,
-					map[string]int{"http-number": 80, "http2-number": 8080}, true, model.Passthrough),
+					map[string][]int{"http-number": {80, 0}, "http2-number": {8080, 0}}, true, model.Passthrough),
```
Comment (bleggett) [inline pilot/pkg/serviceregistry/serviceentry/conversion_test.go:705]: Is the idea here that if 

`servicePort` == `targetPort`, `targetPort` will be 0?

Comment (dddddai) [reply pilot/pkg/serviceregistry/serviceentry/conversion_test.go:705]: `targetPort == 0` means it's unspecified(in the API)

> Should we not just always set it to == `servicePort`?

If so we would not be able to tell if `targetPort` is explicitly set to `servicePort` or unspecified

File: pilot/pkg/serviceregistry/kube/controller/util.go (hunk 1 above)
Comment (dddddai) [reply pilot/pkg/serviceregistry/kube/controller/util.go:114]: `TargetPort` struct contains both a name and a number

For k8s Service:
- targetPort: 8080 -> TargetPort{Name: "", Number: 8080}
- targetPort: "http" -> TargetPort{Name: "http", Number: 0}

For ServiceEntry:
- ```
  ports:
  - number: 80
    name: http
  ```

-> TargetPort{Name: "http", Number: 0}
- ```
  ports:
  - number: 80
    name: http
    targetPort: 8080
  ```
-> TargetPort{Name: "http", Number: 8080}

There should be no functionality loss as long as we follow this semantics

**Review by howardjohn — CHANGES_REQUESTED** [review]
This is very intentionally not a part of the Service object; the targetPort is a property of ServiceInstance, not the Service in our model. Its not clear to me why its being added. How does adding this correlate to fixing the issue linked?

**Comment by dddddai** [comment]
> How does adding this correlate to fixing the issue linked?
https://github.com/istio/istio/blob/a794adc6618b33c9c9f5a98e3cb56a1a2e88f46b/pilot/pkg/serviceregistry/kube/controller/controller.go#L1158-L1171

Updating target ports does not trigger xDS push since they are not included in `model.Service` (`prev.Equals(curr) == true`)

**Comment by dddddai** [comment]
Tried another approach without changing `model.Service`, PTAL

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 3]
```diff
@@ -1138,17 +1138,28 @@ func (c *Controller) servicesForNamespacedName(name types.NamespacedName) []*mod
...
 		return false
 	}
-	return !prev.Equals(curr)
+	if !preConv.Equals(currConv) {
+		return true
+	}
+	if prev != nil && curr != nil {
+		// check if target ports are changed
+		for i := 0; i < len(prev.Spec.Ports); i++ {
+			if prev.Spec.Ports[i].TargetPort != curr.Spec.Ports[i].TargetPort {
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:1159]: could this panic if ports was removed

**Review by hzxuzhonghu — COMMENTED** [review]
I am thinking can we reduce the push scope

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 3 above)
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:1159]: +1. 

Also do we want to compare just targetPort? or the full spec? I guess `conv.Equals` check handles anything else -- but maybe we can comment that in the code?

**Review by howardjohn — COMMENTED** [review]
Nice, I think this version works better

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 3 above)
Comment (dddddai) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:1159]: > could this panic if ports was removed

`preConv.Equals(currConv)` already makes sure the length of ports is not changed

**Comment by hzxuzhonghu** [comment]
>I am thinking can we reduce the push scope

By this, i mean we only need to push to specific proxy, not all proxies

**Review by howardjohn — APPROVED** [review]

**Comment by howardjohn** [comment]
> > I am thinking can we reduce the push scope
> 
> By this, i mean we only need to push to specific proxy, not all proxies

IMO this is not worth the complexity. Its such a rare case that is not worth optimizing, and complex to implement

**Pull request merged**